aiida/cmdline/params/options/config.py

```diff
@@ -25,9 +36,115 @@ def yaml_config_file_provider(handle, cmd_name):  # pylint: disable=unused-argum
     return yaml.safe_load(handle)
 
 
-class ConfigFileOption(OverridableOption):
+def configuration_callback(
+    cmd_name: str | None,
+    option_name: str,
+    config_file_name: str,
+    saved_callback: t.Callable[..., t.Any] | None,
+    provider: t.Callable[..., t.Any],
+    implicit: bool,
+    ctx: click.Context,
+    param: click.Parameter,
+    value: t.Any,
+):
+    """Callback for reading the config file.
+
+    Also takes care of calling user specified custom callback afterwards.
+
+    :param cmd_name: The command name. This is used to determine the configuration directory.
+    :param option_name: The name of the option. This is used for error messages.
+    :param config_file_name: The name of the configuration file.
+    :param saved_callback: User-specified callback to be called later.
+    :param provider: A callable that parses the configuration file and returns a dictionary of the configuration
+        parameters. Will be called as ``provider(file_path, cmd_name)``. Default: ``yaml_config_file_provider``.
+    :param implicit: Whether a implicit value should be applied if no configuration option value was provided.
+    :param ctx: ``click`` context.
+    :param param: ``click`` parameters.
+    :param value: Value passed to the parameter.
+    """
+    ctx.default_map = ctx.default_map or {}
+    cmd_name = cmd_name or ctx.info_name
+
+    if implicit:
+        default_value = os.path.join(click.get_app_dir(cmd_name), config_file_name)  # type: ignore[arg-type]
+        param.default = default_value
+        value = value or default_value
+
+    if value:
+        try:
+            config = provider(value, cmd_name)
+        except Exception as exception:
+            raise click.BadOptionUsage(option_name, f'Error reading configuration file: {exception}', ctx)
+
+        valid_params = [param.name for param in ctx.command.params if param.name != option_name]
+        specified_params = list(config.keys())
+        unknown_params = set(specified_params).difference(set(valid_params))
+
+        if unknown_params:
+            raise click.BadParameter(
+                f'Invalid configuration file, the following keys are not supported: {unknown_params}', ctx, param
```

Comment:
I am open to suggestions for better exceptions but I couldn't find one. Note that for the exception to be properly handled, it needs to be a click exception. I think it makes sense to say there is a problem with the `--config` option, because _its_ value (the content of the file) is incorrect. Since the problem is likely that it contains unsupported keys, we couldn't have it be coupled to any other option of the command, as they don't exist.

Reply:
I'm convinced!